libs/providers/flagd/src/lib/service/in-process/grpc/grpc-fetch.ts

```diff
@@ -35,85 +30,61 @@ export class GrpcFetch implements DataFetch {
+    this._syncStream?.destroy();
     this._syncClient.close();
   }
 
   private listen(
-    resolveConnect: () => void,
     dataFillCallback: (flags: string) => void,
-    connectCallback: () => void,
+    reconnectCallback: () => void,
     changedCallback: (flagsChanged: string[]) => void,
     disconnectCallback: () => void,
+    resolveConnect?: () => void,
+    rejectConnect?: (reason: Error) => void,
   ) {
     this._syncStream = this._syncClient.syncFlags(this._request);
-    this._connecting = false;
 
     this._syncStream.on('data', (data: SyncFlagsResponse) => {
       this._logger?.debug('Received sync payload');
       dataFillCallback(data.flagConfiguration);
-      connectCallback();
       changedCallback([]); // flags changed list not supported
-      resolveConnect();
-      this._nextBackoff = initBackOffMs;
+      // if resolveConnect is undefined, this is a reconnection; we only want to fire the reconnect callback in that case
+      if (resolveConnect) {
+        resolveConnect();
+      } else {
+        reconnectCallback();
+      }
     });
 
     this._syncStream.on('error', (err: ServiceError | undefined) => {
-      if (err?.code === status.CANCELLED) {
-        this._logger?.debug('Stream cancelled, will not be re-established');
-      } else {
-        this._logger?.error('Connection error, attempting to reconnect', err);
-        disconnectCallback();
-        this.reconnectWithBackoff(
-          resolveConnect,
-          dataFillCallback,
-          connectCallback,
-          changedCallback,
-          disconnectCallback,
-        );
-      }
+      this._logger?.error('Connection error, attempting to reconnect', err);
+      disconnectCallback();
+      rejectConnect?.(new GeneralError('Failed to connect stream'));
+      this.reconnect(dataFillCallback, reconnectCallback, changedCallback, disconnectCallback);
     });
   }
 
-  private reconnectWithBackoff(
-    resolver: () => void,
+  private reconnect(
     dataFillCallback: (flags: string) => void,
     reconnectCallback: () => void,
     changedCallback: (flagsChanged: string[]) => void,
     disconnectCallback: () => void,
-  ): void {
-    // avoid reattempts if already connecting
-    // see - https://github.com/grpc/grpc-node/issues/2377
-    if (this._connecting) {
-      return;
-    }
-
-    this._logger?.debug(`Attempting to reconnection after ${this._nextBackoff}ms`);
-    this._connecting = true;
-
-    if (this._nextBackoff > maxStartupDeadlineMs) {
-      resolver();
-    }
-
-    setTimeout(() => {
-      this._nextBackoff = this._nextBackoff * 2;
-      if (this._nextBackoff > maxBackOffMs) {
-        this._nextBackoff = maxBackOffMs;
-      }
-
-      this.listen(resolver, dataFillCallback, reconnectCallback, changedCallback, disconnectCallback);
-    }, this._nextBackoff);
+  ) {
+    const channel = this._syncClient.getChannel();
+    channel.watchConnectivityState(channel.getConnectivityState(true), Infinity, () => {
```

Comment:
I don't think we need to get connectivity state with connect retry set to true. This is only needed if the connection is in idle state. In reconnect, we should see a transition from `READY` to `TRANSIENT_FAILURE`

Besides, do we need this as a callback from `channel.watchConnectivityState` 🤔  reason I am asking, we only end up in reconnect only for error. Which means we anyway need to call for a reconnect.

Reply:
I took this pattern based on a comment from one of the gRPC-js maintainers here: https://github.com/grpc/grpc-node/issues/2218#issuecomment-1240994000

It seems this is a good method for reacting to channel readiness.